src/runtime_src/core/include/xrt/xrt_pskernel.h

```diff
@@ -0,0 +1,713 @@
+  /// @endcond
+
+private:
+  std::shared_ptr<psrun_impl> handle;
+
+  XCL_DRIVER_DLLESPEC
+  void
+  set_arg_at_index(int index, const void* value, size_t bytes);
+
+  XCL_DRIVER_DLLESPEC
+  void
+  set_arg_at_index(int index, const xrt::bo&);
+
+  template<typename ArgType, typename ...Args>
+  void
+  set_arg(int argno, ArgType&& arg, Args&&... args)
+  {
+    set_arg(argno, std::forward<ArgType>(arg));
+    set_arg(++argno, std::forward<Args>(args)...);
+  }
+};
+ 
+
+/*!
+ * @class kernel
+ *
+ * A kernel object represents a set of instances matching a specified name.
+ * The kernel is created by finding matching kernel instances in the 
+ * currently loaded xclbin.
+ *
+ * Most interaction with kernel objects are through \ref xrt::run objects created 
+ * from the kernel object to represent an execution of the kernel
+ */
+class pskernel_impl;
+class pskernel
+{
+public:
+  /**
+   * cu_access_mode - compute unit access mode
+   *
+   * @var shared
+   *  CUs can be shared between processes
+   * @var exclusive
+   *  CUs are owned exclusively by this process
+   */
+  enum class cu_access_mode : uint8_t { exclusive = 0, shared = 1, none = 2 };
+
+  /**
+   * pskernel() - Construct for empty kernel
+   */
+  pskernel() = default;
+
+  /**
+   * pskernel() - Constructor from a device and xclbin
+   *
+   * @param device
+   *  Device on which the kernel should execute
+   * @param xclbin_id
+   *  UUID of the xclbin with the kernel
+   * @param name
+   *  Name of kernel to construct
+   * @param mode
+   *  Open the kernel instances with specified access (default shared)
+   *
+   * The kernel name must uniquely identify compatible kernel
+   * instances (compute units).  Optionally specify which kernel
+   * instance(s) to open using
+   * "kernelname:{instancename1,instancename2,...}" syntax.  The
+   * compute units are default opened with shared access, meaning that
+   * other kernels and other process will have shared access to same
+   * compute units.
+   */
+  XCL_DRIVER_DLLESPEC
+    pskernel(const xrt::device& device, const xrt::uuid& xclbin_id, const std::string& name,
+         cu_access_mode mode = cu_access_mode::shared);
+
+  /// @cond
+  /// Deprecated construtor for exclusive access
+  pskernel(const xrt::device& device, const xrt::uuid& xclbin_id, const std::string& name, bool ex)
+    : pskernel(device, xclbin_id, name, ex ? cu_access_mode::exclusive : cu_access_mode::shared)
+  {}
+
+  /**
+   * Obsoleted construction from xclDeviceHandle
+   */
+  XCL_DRIVER_DLLESPEC
+  pskernel(xclDeviceHandle dhdl, const xrt::uuid& xclbin_id, const std::string& name,
+         cu_access_mode mode = cu_access_mode::shared);
+  /// @endcond
+
+  /**
+   * operator() - Invoke the kernel function
+   *
+   * @param args
+   *  Kernel arguments
+   * @return 
+   *  Run object representing this kernel function invocation
+   */
+  template<typename ...Args>
+  psrun
+  operator() (Args&&... args)
+  {
+    psrun r(*this);
```

Comment:
```suggestion
    psrun r { *this };
```
would be less cryptic compared to the following line